packages/core/src/PartialExecutionUtils/recreateNodeExecutionStack.ts

```diff
@@ -0,0 +1,166 @@
+import {
+	NodeConnectionType,
+	type IExecuteData,
+	type INode,
+	type INodeExecutionData,
+	type IPinData,
+	type IRunData,
+	type ISourceData,
+	type ITaskDataConnectionsSource,
+	type IWaitingForExecution,
+	type IWaitingForExecutionSource,
+} from 'n8n-workflow';
+
+import * as a from 'assert/strict';
+import type { DirectedGraph } from './DirectedGraph';
+import { getIncomingData } from './getIncomingData';
+import { getSourceDataGroups } from './getSourceDataGroups';
+
+export function recreateNodeExecutionStack(
+	graph: DirectedGraph,
+	startNodes: INode[],
+	destinationNode: INode,
+	runData: IRunData,
+	pinData: IPinData,
+): {
+	nodeExecutionStack: IExecuteData[];
+	waitingExecution: IWaitingForExecution;
+	waitingExecutionSource: IWaitingForExecutionSource;
+} {
+	// Validate invariants.
```

Comment:
I'm rewriting this function in my next PR: https://github.com/n8n-io/n8n/pull/10758/files#diff-b37ed612d467d57f6089dce0fce290530a0a13694b054db61132b18d1cd25eaf

I can make those changes there. I left a todo for me there already because this function becomes overwhelming:
https://github.com/n8n-io/n8n/pull/10758/files#diff-b37ed612d467d57f6089dce0fce290530a0a13694b054db61132b18d1cd25eafR58